content/articles/sharing-deployments-with-nixops.md

```diff
@@ -0,0 +1,198 @@
+with some defaults such as absolute paths make it a bit cumbersome to
+share deployments. This post describes how to work around these
+issues.
+
+<!-- PELICAN_END_SUMMARY -->
+
+## Defining a server
+
+We'll define an extremely simple server that doesn't do anything other
+than warming this planet:
+
+```nix
+# network.nix
+let
+    region = "eu-west-1";
+in rec {
+    network.description = "test-network";
+    resources.ec2KeyPairs.test-keys = { inherit region; };
+
+    server = { resources, pkgs, lib, ... }: {
+        deployment.targetEnv = "ec2";
+        deployment.ec2.region = region;
+        deployment.ec2.instanceType = "t1.micro";
+        deployment.ec2.spotInstancePrice = 7;
+        deployment.ec2.keyPair = resources.ec2KeyPairs.test-keys;
+    };
+}
+```
+
+
+### Trick 1 and 2: Confining NixOps to a single directory
+
+Our first two tricks are to circumvent the default NixOps-behaviour of
+using absolute paths (by using angular brackets `<...>`), and to store
+the state for our deployment in a local file (by using `-s ...`):
+
+```
+$ NIX_PATH=${NIX_PATH}:. nixops create -s localstate.nixops -d test '<network.nix>'
+created deployment ‘1b699a42-5709-11e5-ad66-542696dce997’
+1b699a42-5709-11e5-ad66-542696dce997
+```
+
+The angular brackets `<...>` tell the nix command to search for
+`network.nix` in the `NIX_PATH`. And we just set `NIX_PATH` to point
+to the current directory.
+
+Using a local state file (`-s localstate.nix`) means we can check the
+state into git (more on that later).
+
+
+### Trick 3: Using direnv to simplify
+
+That's a lot of environment variables and parameters to type. In order
+to clean up the command invocation we're using a wonderful tool called
+[direnv](https://github.com/direnv/direnv) which exports environment
+variables from a `.envrc` file when entering a directory.
+
+An `.envrc` for AWS deployment typically looks like this:
+
+```bash
+export AWS_ACCESS_KEY_ID="AKIA...."
+export AWS_SECRET_ACCESS_KEY="eexie9Ieeel9XieCThish5..."
+export NIX_PATH=${NIX_PATH}:.
+export NIXOPS_STATE=localstate.nixops
+```
+
+We can now inspect the test network we just created with `nixops
+info`:
+
+```console
+$ nixops info
+Network name: test
+Network UUID: f29084f8-570c-11e5-83af-542696dce997
+Network description: test-network
+Nix expressions: <network.nix>
+
+[...]
+```
+
+And deploy:
+
+```console
+$ nixops deploy
+test-keys> uploading EC2 key pair ‘charon-f29084f8-570c-11e5-83af-542696dce997-test-keys’...
+server...> creating EC2 instance (AMI ‘ami-0126a576’, type ‘t1.micro’, region ‘eu-west-1’)...
+server...> waiting for spot instance request ‘sir-01pq1vda’ to be fulfilled... [pending-evaluation] [fulfilled]
+server...> cancelling spot instance request ‘sir-01pq1vda’... [request-canceled-and-instance-running]
+server...> waiting for IP address... [pending] [running] 54.76.8.50 / 172.31.30.116
+server...> waiting for SSH......
+```
+
+
+### Trick 4: Sharing secrets in git
```

Comment:
I tried to fit that in but could not find a succinct way of going into the full key distribution. If you have any ideas let me know!